### PR TITLE
Callback security and usr

### PR DESCRIPTION
**Pull request by Cyberboss**
Callbacks now preserve usr of the creator if usr is null at the time of it's invocation (timers)
var_edited is now set on admin created datums

### CODE REVIEW

**Comment by RemieRichards** [comment]
```var_edited is now set on admin created datums```
Dumb.

File: code/datums/callback.dm [hunk 1]
```diff
@@ -70,8 +73,16 @@
 		call(thingtocall, proctocall)(arglist(calling_arguments))
 
 /datum/callback/proc/Invoke(...)
+	if(!usr)
+		var/datum/weakref/W = user
+		if(W)
+			var/mob/M = W.resolve()
+			if(M)
+				return world.PushUsr(user, src)
```
Comment (duncathan) [inline code/datums/callback.dm]: shouldn't you be passing `M` instead of `user`?

File: code/datums/callback.dm [hunk 2]
```diff
@@ -87,8 +98,17 @@
 //copy and pasted because fuck proc overhead
 /datum/callback/proc/InvokeAsync(...)
 	set waitfor = FALSE
+
+	if(!usr)
+		var/datum/weakref/W = user
+		if(W)
+			var/mob/M = W.resolve()
+			if(M)
+				return world.PushUsr(user, src)
```
Comment (duncathan) [inline code/datums/callback.dm]: same here

**Review by duncathan — CHANGES_REQUESTED** [review]
I also don't understand the point of setting `var_edited` for new datums until a var is actually edited. Could you explain that?

**Comment by Cyberboss** [comment]
There's an exploit with admin created callbacks that can be used to get around the admin proc call protections

**Review by duncathan — APPROVED** [review]

**Pull request merged**